### ISSUE TITLE
Improve SemanticdbIndex

### ISSUE THREAD

**Issue by olafurpg** [scalacenter/scalafix#704]
The current `SemanticdbIndex` trait has problems:

* It exposes obsolete `scala.meta.{Denotation,Database,Document,Symbol}`
* It's wasteful with memory for no good reason
* It's slow to load
* It's still too low-level for end-users, who care primarily about finding types and symbols of tree nodes
* The trait has only one implementation `EagerSemanticdbIndex`, and the codebase currently downcasts in several places (https://github.com/scalacenter/scalafix/pull/699/files#r181519292)

This ticket is to discuss the design of a new API with the following goals

* expose as few data structures as possible, the exposed data structures should live in the scalafix repo
* low memory pressure by separating index per source file
* no more downcasts to single trait implementation
* focus on core operations: `Tree.symbol: Symbol` and `Symbol.info: SymbolInformation`

**Comment by olafurpg** [comment]
Here is a concrete proposal:

```scala
// before
case class MyRule(index: SemanticdbIndex) extends SemanticRule(index, "MyRule") {
  def fix(ctx: RuleCtx): Patch
}

// after
case object MyRule extends SemanticRule("MyRule") {
  def fix(implicit doc: SemanticDoc): Patch
}
```

I would also like to rename `RuleCtx` into `doc: Document` for syntactic files (only `.tree/.tokens`) and then use `doc: SemanticdbDocument` for semantic files (syntactic + `.symbol`).
In comparison with `SemanticdbIndex`, `SemanticDoc` will be different in the following ways

- No `.database: Database`
- No `.documents: Seq[Document]`
- `.messages` is deprecated, but aliases `.diagnostics` which returns a source compatible API as before
- `.symbol(Position): Option[Symbol]` and `.symbol(Tree): Option[Symbol]` are unchanged
- `.sym(Position): Sym` is a new addition, where `Sym` is a wrapper around a string.
- `.denotation` is deprecated, but  aliases `.info: Sym.Info` which is compatible with old `Denotation` but contains deprecated methods that alias new `.kind/.props` methods which correspond to the SemanticDB v3 spec.
- No `.withDocument` 

It should be possible to provide a scalafix migration rewrite for the source breaking changes but the migration from `.symbol` to `.sym` and `.denotation` to `.info` is probably best to do manually. There are no changes to the existing `RuleCtx` or `Patch` API.

I am open for alternative suggestions for the (slightly cryptic) name `Sym`, but unfortunately it can't be `Symbol` since that is already taken by `scala.meta.Symbol`. In essence, `Sym` is a thin wrapper around a string with the following API
```scala
final class Sym private (val value: String) {
  def isNone: Boolean
  def isRootPackage: Boolean
  def isEmptyPackage: Boolean
  def isGlobal: Boolean
  def isLocal: Boolean
  def owner: Sym
  def info(doc: SemanticDoc): Sym.Info
}
```


And `Sym.Info` is defined as follows inside of the `Sym` companion object
```scala
final class Info private[scalafix] (info: s.SymbolInformation) {
  def sym: Sym = new Sym(info.symbol)
  def owner: Sym = new Sym(info.owner)
  def name: String = info.name
  def kind: Kind = new Kind(info)
  def props: Properties = new Properties(info.properties)
  def access: Accessibility =
    new Accessibility(info.accessibility.getOrElse(s.Accessibility()))

  // privates
  private[scalafix] def tpe: s.Type =
    info.tpe.getOrElse(s.Type())
}

final class Kind private[Sym] (info: s.SymbolInformation) {
  def isClass: Boolean = k.isClass
  def isObject: Boolean = k.isObject
  def isTrait: Boolean = k.isTrait
  def isDef: Boolean = k.isMethod && !is(p.VAL) && !is(p.VAR)
  def isVal: Boolean = k.isMethod && is(p.VAL)
  def isVar: Boolean = k.isMethod && is(p.VAR) && !isSetter
  def isVarSetter: Boolean = k.isMethod && is(p.VAR) && isSetter
  def isMacro: Boolean = k.isMacro
  def isConstructor: Boolean = k.isConstructor
  def isType: Boolean = k.isType
  def isParameter: Boolean = k.isParameter
  def isTypeParameter: Boolean = k.isTypeParameter
  def isPackage: Boolean = k.isPackage
  def isPackageObject: Boolean = k.isPackageObject

  // privates
  private[this] def k = info.kind
  private[this] def is(property: s.SymbolInformation.Property): Boolean =
    (info.properties & property.value) != 0
  private[this] def isSetter = info.name.endsWith("_=")
}

final class Properties private[Sym] (props: Int) {
  def isAbstract: Boolean = is(p.ABSTRACT)
  def isFinal: Boolean = is(p.FINAL)
  def isSealed: Boolean = is(p.SEALED)
  def isImplicit: Boolean = is(p.IMPLICIT)
  def isLazy: Boolean = is(p.LAZY)
  def isCase: Boolean = is(p.CASE)
  def isCovariant: Boolean = is(p.COVARIANT)
  def isContravariant: Boolean = is(p.CONTRAVARIANT)
  def isStatic: Boolean = is(p.STATIC)
  def isPrimary: Boolean = is(p.PRIMARY)
  def isEnum: Boolean = is(p.ENUM)

  // privates
  private[this] def is(property: s.SymbolInformation.Property): Boolean =
    (props & property.value) != 0
}

final class Accessibility private[Sym] (a: s.Accessibility) {
  def isPrivate: Boolean = a.tag.isPrivate
  def isPrivateThis: Boolean = a.tag.isPrivateThis
  def isProtected: Boolean = a.tag.isProtected
  def isProtectedThis: Boolean = a.tag.isProtectedThis
  def isPublic: Boolean = a.tag.isPublic
  def within: Option[Sym] =
    if (a.symbol.isEmpty) scala.None
    else Some(new Sym(a.symbol))
}

```

**Comment by xeno-by** [comment]
@olafurpg Thanks for the heads up! Do you by any chance have a compilable sketch of the API? I have some questions, but many of them would be directly answered by code.

**Comment by olafurpg** [comment]
I'll try to implement this proposal today.

**Comment by olafurpg** [comment]
Still working on this. It's not an easy refactor so I'm looking at ways to write the new API independently of the old `SemanticdbIndex` and `RuleCtx` API and then provide converters to the old API to ease migration.

**Comment by xeno-by** [comment]
@olafurpg I left a review for the API portion of the PR. It would also help to get a milestone release once you feel that `v1` is ready for a test drive - that way we'll have easier time providing feedback.

**Comment by olafurpg** [comment]
Merged in #737, there is still a lot of work to iterate on the new api in `scalafix.v1` but those discussions can be moved to a separate ticket.
